### PR TITLE
Support parsing Exception section

### PR DESCRIPTION
**Pull request by bkotsopoulossc**
Issue: https://github.com/GoogleChromeLabs/wasmbin/issues/7

Testing (since the original proposal called the new section Tag, wasm-objdump still uses this nomenclature):

```
% cat hello-world.cc
#include <iostream>

int main(int argc, char** argv) {
  if (argc > 3) {
    throw std::runtime_error("some err");
  }
  std::cout << "hello world!" << std::endl;
  return 0;
}
% emcc -fwasm-exceptions -g hello-world.cc
% wasm-objdump -h a.out.wasm          

a.out.wasm:	file format wasm 0x1

Sections:

     Type start=0x0000000e end=0x00000200 (size=0x000001f2) count: 62
   Import start=0x00000206 end=0x00000358 (size=0x00000152) count: 12
 Function start=0x0000035e end=0x000009b4 (size=0x00000656) count: 1620
    Table start=0x000009ba end=0x000009c1 (size=0x00000007) count: 1
   Memory start=0x000009c7 end=0x000009cd (size=0x00000006) count: 1
      Tag start=0x000009d3 end=0x000009d6 (size=0x00000003) count: 1
   Global start=0x000009dc end=0x000009ef (size=0x00000013) count: 3
   Export start=0x000009f5 end=0x00000b49 (size=0x00000154) count: 20
     Elem start=0x00000b4f end=0x00000e3d (size=0x000002ee) count: 1
     Code start=0x00000e43 end=0x00025b63 (size=0x00024d20) count: 1620
     Data start=0x00025b69 end=0x00029c89 (size=0x00004120) count: 2
   Custom start=0x00029c8f end=0x00054377 (size=0x0002a6e8) "name"
   Custom start=0x0005437d end=0x0012ae57 (size=0x000d6ada) ".debug_info"
   Custom start=0x0012ae5d end=0x00164a31 (size=0x00039bd4) ".debug_loc"
   Custom start=0x00164a37 end=0x0016e545 (size=0x00009b0e) ".debug_ranges"
   Custom start=0x0016e54b end=0x0017ec1b (size=0x000106d0) ".debug_abbrev"
   Custom start=0x0017ec21 end=0x001d5996 (size=0x00056d75) ".debug_line"
   Custom start=0x001d599c end=0x00239f09 (size=0x0006456d) ".debug_str"
```

Successfully able to parse with
```
Module::decode_from(BufReader::new(File::open("a.out.wasm")?))?;
```

And support in the dump example:
```
% ./target/release/examples/dump ~/Snapchat/Dev/wasm-testing/exceptions.wasm exception
Exception(
    Blob(Output {
        value: [
            Exception {
                type_index: Type#4,
            },
        ],
    }),
)
Found 1 sections.
```

PTAL @RReverser

### CODE REVIEW

**Comment by RReverser** [comment]
You might also want to look into replacing numeric props with something more specialised. I'm on the phone, can't look them up right now, but e.g. shouldn't  `ty` be a `TypeId`?

**Comment by RReverser** [comment]
Yup, you can create an enum with one variant for now - that would be future-proof and easy to ignore for those who don't need it.

**Comment by RReverser** [comment]
One more thing - you've marked this as closing #7, but I thought the issue is about complete exceptions support. Would you want to add support for try-catch (and other relevant structures if any) as part of the same PR or separately?

**Comment by bkotsopoulossc** [comment]
Yeah that’s a good point. There’s still definitely more work to fully support exceptions. There’s the new type for the import section and export section and the new instructions like you mentioned. And probably gating all of this behind a feature flag like you did with simd and threads. And updating the test suite to test it all.

For my use case this was enough, because all the individual sections are parsed lazily, so I just needed to be able to hand the tag section as that’s parsed up front. I don’t have time at the moment to go further but I’m happy to leave the issue open or rename it

**Comment by RReverser** [comment]
FWIW when looking at proposals, it's better not to look at overview Markdown files - those are usually created in early stages and remain outdated - but rather at the actual rendered spec.

I've just looked at the one for exception handling and it looks like the section ended up being called "Exception section" after all: https://webassembly.github.io/exception-handling/core/binary/modules.html#exception-section

It's also not clear if `tag_type` structure and `attrs` field in the Markdown are actually related to each other, and the binary spec simply specifies there is a 0x00 byte before the type id. In that case, perhaps it's best for us to skip that field for now and use the `discriminant` to account for that byte. You can see an example of how it's done in the existing `FuncType` struct.

**Comment by bkotsopoulossc** [comment]
> FWIW when looking at proposals, it's better not to look at overview Markdown files - those are usually created in early stages and remain outdated - but rather at the actual rendered spec.

I didn't realize that, thanks. Unfortunately those are a lot harder for me to understand than the markdown :/

> I've just looked at the one for exception handling and it looks like the section ended up being called "Exception section" after all: https://webassembly.github.io/exception-handling/core/binary/modules.html#exception-section

> It's also not clear if tag_type structure and attrs field in the Markdown are actually related to each other, and the binary spec simply specifies there is a 0x00 byte before the type id. In that case, perhaps it's best for us to skip that field for now and use the discriminant to account for that byte. You can see an example of how it's done in the existing FuncType struct.

This does seem problematic. To add to this confusion, I'm not sure if this latest version of the spec is supported at the framework level yet. Which makes it hard to test and hard to land this change if it's not actually useful yet in practice.

This makes me wonder if the right approach here is to instead allow unknown discriminants in the module section parsing to allow for future sections, and let them pass through untouched when re-serializing...

**Comment by RReverser** [comment]
> This does seem problematic. To add to this confusion, I'm not sure if this latest version of the spec is supported at the framework level yet. Which makes it hard to test and hard to land this change if it's not actually useful yet in practice.

Binary-wise it appears to be exactly the same, it's just that the binary spec doesn't assign any meaning to that same 0x00 byte.

> Unfortunately those are a lot harder for me to understand than the markdown :/

Oh interesting. I usually found them a lot cleaner, containing only the info necessary for the parsing - that's how I described most of the data structures in this crate.

**Comment by bkotsopoulossc** [comment]
> Oh interesting. I usually found them a lot cleaner, containing only the info necessary for the parsing - that's how I described most of the data structures in this crate.

I guess I'm mainly confused because [this section](https://webassembly.github.io/exception-handling/core/binary/modules.html#exception-section) says an `exn = {0x00, typeidx}` and [this section](https://webassembly.github.io/exception-handling/core/syntax/modules.html#syntax-exn) says `exn = {typeidx}`

**Comment by RReverser** [comment]
With my suggestion above, the result should look something like

```rust
#[wasmbin(discriminant = 0x00)]
pub struct Exception {
    pub ty: TypeId,
}
```

**Comment by bkotsopoulossc** [comment]
@RReverser is this alright?

File: src/sections.rs [hunk 1]
```diff
@@ -242,6 +240,13 @@ pub struct Locals {
     pub ty: ValueType,
 }
 
+// https://webassembly.github.io/exception-handling/core/binary/modules.html#exception-section
+#[derive(Wasmbin, WasmbinCountable, Debug, Arbitrary, PartialEq, Eq, Hash, Clone, Visit)]
+#[wasmbin(discriminant = 0x00)]
+pub struct Exception {
+    pub type_index: TypeId,
```
Comment (RReverser) [inline src/sections.rs:247]: Nit: I'd go for a common `ty` or at least `type_id`.

**Comment by RReverser** [comment]
By the way, this is not a stable part of the Wasm spec, so please define a feature flag in Cargo.toml and use it to conditionally compile exception paths: https://github.com/GoogleChromeLabs/wasmbin/blob/master/Cargo.toml#L31

**Comment by RReverser** [comment]
I rebased & merged #11 on top of this one and will try to address the remaining changes to get it all in.

**Comment by bkotsopoulossc** [comment]
@RReverser thanks a bunch, sorry I wasn't able to get to it :/

**Pull request merged**